File: chap4/sec4.4/exampleSTMG.m
l=7;                                             % with the choice l=3 and
J=2^l-1;                                         % N=(J+1)*2^(3*l)-1 the 
N=J;                                             % space coarsening condition 
e=ones(J,1);                                     % fails, try it 
dx=1/(J+1); x=(0:dx:1)';
A=1/dx^2*spdiags([e -2*e e],[-1 0 1],J,J);
T=5; dt=T/N; t=(0:dt:T);
u0=@(x) zeros(size(x));
gl=@(t) zeros(size(t)); gr=@(t) zeros(size(t));
f=@(x,t) x.^4.*(1-x).^4+10*sin(8*t);
for n=1:N                                        % compute rhs b for reuse
  b(:,n)=dt*feval(f,x(2:end-1),t(n+1));          % source function
  b(1,n)=b(1,n)+dt/dx^2*gl(t(n+1));              % boundary conditions
  b(end,n)=b(end,n)+dt/dx^2*gr(t(n+1));
end;

u(:,1)=u0(x); u(1,1:N+1)=gl(t); u(end,1:N+1)=gr(t);
G=speye(size(A))-dt*A;
for n=1:N                                        % compute exact solution
  u(2:end-1,n+1)=G\(u(2:end-1,n)+b(:,n));        % exact BE
end
uBE=u;

Jc=(J+1)/2-1;                                    % coarse grid size
P=sparse(J,Jc);                                  % prolongation by interpolation
for j=1:Jc
  P(2*j,j)=1; P(2*j-1,j)=0.5; P(2*j+1,j)=0.5;
end;
R=0.5*P';                                        % restriction
Ac=R*A*P;                                        % coarse matrix by Galerkin 
Nc=(N+1)/2-1;
Pt=sparse(N,Nc);                                 % prolongation by interpolation
for j=1:Nc
  Pt(2*j,j)=1; Pt(2*j-1,j)=0.5; Pt(2*j+1,j)=0.5;
end;
Rt=0.5*Pt';                                      % restriction
Pt(end,end)=1;                                   % no final zero bc in time
Gc=speye(size(Ac))-2*dt*Ac;                      % coarsening in time also

rng('default')                                   % random initial guess
u(2:end-1,2:end)=rand(J,N);                      % with correct ic and bc
errcxt(1)=max(max(abs(uBE-u)));
for k=1:10
  for j=1:nu                                     % use nu block Jacobi steps
    uo=u;                                        % block Jacobi is parallel
    for n=1:N
      u(2:end-1,n+1)=(1-al)*uo(2:end-1,n+1)+al*(G\(uo(2:end-1,n)+b(:,n)));
    end;
  end
  mesh(x,t,uBE'-u'); xlabel('x'); ylabel('t');
  title('Error after presmoothing');
  pause
  for n=1:N                                      % compute residual
    r(:,n)=u(2:end-1,n)+b(:,n)-G*u(2:end-1,n+1);
  end;
  rc=R*r;                                        % restrict residual in space
  rc=rc*Rt';                                     % restrict residual in time 
  uc=zeros(Jc+2,1);                              % zero ic for correction
  for n=1:Nc                                     % coarse correction
    uc(2:end-1,n+1)=Gc\(uc(2:end-1,n)+rc(:,n));  % exact BE using 2*dt
  end                                            % extend in space and time
  u(2:end-1,2:end)=u(2:end-1,2:end)+P*uc(2:end-1,2:end)*Pt';
  mesh(x,t,uBE'-u'); xlabel('x'); ylabel('t');
  title('Error after coarse correction');
  pause
  for j=1:nu                                     % use nu block Jacobi steps
    uo=u;                               
    for n=1:N
      u(2:end-1,n+1)=(1-al)*uo(2:end-1,n+1)+al*(G\(uo(2:end-1,n)+b(:,n)));
    end;
  end
  errcxt(k+1)=max(max(abs(uBE-u)));
  mesh(x,t,uBE'-u'); xlabel('x'); ylabel('t');
  title(['Error after 2 grid iteration k=' num2str(k)]);
  pause
end

File: chap4/sec4.4/exampleSTMG.py
import numpy as np
import scipy.sparse as sp
import scipy.sparse.linalg as spla
import matplotlib.pyplot as plt

# Parameters
l = 7
J = 2**l - 1
N = J
e = np.ones(J)
dx = 1 / (J + 1)
x = np.linspace(0, 1, J+2)

# Matrix A
diagonals = np.array([e, -2*e, e]) / dx**2
A = sp.spdiags(diagonals, [-1, 0, 1], J, J)

# Time grid
T = 5
dt = T / N
t = np.linspace(0, T, N+1)

# Initial and boundary conditions
u0 = lambda x: np.zeros_like(x)
gl = lambda t: np.zeros_like(t)
gr = lambda t: np.zeros_like(t)
f = lambda x, t: x**4 * (1-x)**4 + 10 * np.sin(8*t)

# Compute rhs b for reuse
b = np.zeros((J, N))
for n in range(N):
    b[:, n] = dt * f(x[1:-1], t[n+1])
    b[0, n] += dt / dx**2 * gl(t[n+1])
    b[-1, n] += dt / dx**2 * gr(t[n+1])

# Initial solution
u = np.zeros((J+2, N+1))
u[:, 0] = u0(x)
u[0, :] = gl(t)
u[-1, :] = gr(t)

# BE time stepping matrix
G = sp.eye(J) - dt * A

# Compute exact solution
for n in range(N):
    u[1:-1, n+1] = spla.spsolve(G, u[1:-1, n] + b[:, n])

uBE = u.copy()

# Coarse grid setup in space
Jc = (J + 1) // 2 - 1
P = sp.lil_matrix((J, Jc))
for j in range(Jc):
    P[2*j+1, j] = 1
    P[2*j, j] = 0.5
    P[2*j + 2, j] = 0.5

P = P.tocsc()
R = 0.5 * P.T
Ac = R @ A @ P

# Coarse grid setup in time
Nc = (N + 1) // 2 - 1
Pt = sp.lil_matrix((N, Nc))
for j in range(Nc):
    Pt[2*j+1, j] = 1
    Pt[2*j, j] = 0.5
    Pt[2*j + 2, j] = 0.5

Pt = Pt.tocsc()
Rt = 0.5 * Pt.T
Pt[-1, -1] = 1  # No final zero bc in time
Gc = sp.eye(Jc) - 2 * dt * Ac

# Multigrid iteration
nu = 5
al = 0.5
np.random.seed(0)
u[1:-1, 1:] = np.random.rand(J, N)
errcxt = np.zeros(11)
errcxt[0] = np.max(np.abs(uBE - u))

fig = plt.figure("exampleParabolicMultigrid", layout='tight')
ax = fig.add_subplot(111, projection='3d')

for k in range(10):
    for _ in range(nu):  # Block Jacobi steps
        uo = u.copy()
        for n in range(N):
            u[1:-1, n+1] = (1-al) * uo[1:-1, n+1] + al * spla.spsolve(G, uo[1:-1, n] + b[:, n])

    # Plot error after presmoothing
    X, T = np.meshgrid(x, t)

    ax.cla()
    ax.plot_surface(X, T, (uBE - u).T, cmap='viridis',
                    rstride=1, cstride=1, shade=False)
    ax.set_xlabel('x')
    ax.set_ylabel('t')
    ax.set_title('Error after presmoothing')
    plt.pause(0.5)

    # Compute residual
    r = np.zeros((J, N))
    for n in range(N):
        r[:, n] = u[1:-1, n] + b[:, n] - G * u[1:-1, n+1]

    # Restrict residual in space and time
    rc = R * r
    rc = rc @ Rt.T

    # Coarse correction
    uc = np.zeros((Jc+2, Nc+1))
    for n in range(Nc):
        uc[1:-1, n+1] = spla.spsolve(Gc, uc[1:-1, n] + rc[:, n])

    # Extend in space and time
    u[1:-1, 1:] += P @ uc[1:-1, 1:] @ Pt.T

    # Plot error after coarse correction
    ax.cla()
    ax.plot_surface(X, T, (uBE - u).T, cmap='viridis',
                    rstride=1, cstride=1, shade=False)
    ax.set_xlabel('x')
    ax.set_ylabel('t')
    ax.set_title('Error after coarse correction')
    plt.pause(0.5)

    for _ in range(nu):  # Block Jacobi steps
        uo = u.copy()
        for n in range(N):
            u[1:-1, n+1] = (1-al) * uo[1:-1, n+1] + al * spla.spsolve(G, uo[1:-1, n] + b[:, n])

    errcxt[k+1] = np.max(np.abs(uBE - u))

    # Plot error after 2-grid iteration
    ax.cla()
    ax.plot_surface(X, T, (uBE - u).T, cmap='viridis',
                    rstride=1, cstride=1, shade=False)
    ax.set_xlabel('x')
    ax.set_ylabel('t')
    ax.set_title(f'Error after 2-grid iteration k={k}')
    plt.pause(0.5)
